Step1 - 구간 추가 요구사항 반영

seonje1 opened the pull request:
안녕하세요 1단계 미션 PR입니다.

seonje1 committed:
인수 테스트 시나리오 작성

seonje1 committed:
구간 등록 추가 요청사항 기능 구현

seonje1 committed:
0단계 피드백 반영

seonje1 committed:
인수 테스트 시나리오 추가 - 계산한 거리가 1보다 작은 경우 예외발생

testrace commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -54,4 +55,16 @@ public Station getDownStation() {
         return line;
     }
+
+    public void changeUpStation(Station upStation, int distance) {
+        this.upStation = upStation;
+        renewDistance(this.distance - distance);
+    }
```
단위 테스트를 추가하면 어떨까요?

testrace commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -20,46 +21,55 @@ public Sections() {
 
     public List<Section> getSections() {
-        return sections;
+        return sections.stream()
+                .sorted((s1, s2) -> {
+                    if (s1.getDownStation().equals(s2.getUpStation())) {
+                        return -1;
+                    }
+                    return 0;
+                })
+                .collect(Collectors.toList());
```
추가된 모든 기능에 단위 테스트가 없는데요
미션 주제에 맞게 인수 테스트 작성 이후 기능 구현 중에 TDD 사이클로 구현해 보시면 좋을 것 같아요

testrace commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -20,46 +21,55 @@ public Sections() {
 
     public List<Station> getStations() {
-        List<Station> stations = sections.stream().map(Section::getUpStation).collect(Collectors.toList());
+        List<Station> stations = getSections().stream().map(Section::getUpStation).collect(Collectors.toList());
```
개행으로 가독성을 높이면 어떨까요?
IntelliJ를 사용하신다면 각 메서드마다의 리턴 타입도 확인할 수 있어요
```suggestion
        List<Station> stations = getSections().stream()
            .map(Section::getUpStation)
            .collect(Collectors.toList());
```

testrace commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -20,46 +21,55 @@ public Sections() {
         }
     }
 
+    private void sortSections(Section section) {
```
메서드명에 신규 구간을 등록하기 위해 기준이 되는 구간을 수정한다는 의도를 드러내면 어떨까요?
메서드명을 봤을 땐 sections 중 하나를 수정(chageUpStation 메서드 호출)할 것이라고 예상하기 어려운 것 같아요

testrace commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -20,46 +21,55 @@ public Sections() {
 
+    private void sortSections(Section section) {
+        sections.stream()
+                .filter(s -> s.getUpStation() == section.getUpStation())
```
객체간의 비교는 `equals` 메서드를 활용하면 어떨까요?

testrace commented on src/test/java/nextstep/subway/acceptance/section/SectionAcceptanceTest.java:
```diff
@@ -93,6 +114,23 @@ void generateAlreadySection() {
+        //then
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+        assertThat(response.asPrettyString()).isEqualTo("거리가 1보다 작을 수 없습니다.");
+    }
```
예외 케이스 인수 테스트를 만들어 주셨군요 👍 
인수 조건의 '계산한 거리가 1보다 작은 경우'도 좋지만 무엇을 어떻게 계산했는지 파악하기 쉽지 않은 것 같아요
기존 노선의 거리보다 긴 구간을 추가 할 수 없다는 인수 조건으로 대체하면 어떨까요?

testrace reviewed (changes requested):
안녕하세요 선제님 😃 
1단계 깔끔하게 잘 구현해 주셨네요 👍 
몇 가지 코멘트 남겨두었는데 확인해 주시고 다시 리뷰 요청해 주세요

seonje1 committed:
피드백 수정사항 반영

- 구간 조회 정렬 기능 추가 단위테스트 작성
- 노선 가운데 구간 등록 기능, 길이에 따른 예외케이스 단위테스트 작성
- 메소드 네이밍 수정 sortSection -> middleSectionAddForPreviousUpStationChange
- 인수테스트 시나리오 수정 계산한 거리가 1보다 작은 경우 -> 기존 구간의 거리보다 등록할 구간의 거리가 긴 경우
- stream식 개행

seonje1 committed:
사용하지않는 주석 제거

testrace reviewed (approved):
안녕하세요 선제님 😃 
피드백 잘 반영해 주셨네요 👍 
다음 단계 진행해 주세요!
